### PR TITLE
[Megatron-FSDP] Add conditional param.grad dereferencing logic to support full-iteration (FWD-BWD) CUDA graphability.

### PR DESCRIPTION
**Pull request by cspades**
# What does this PR do ?

- Fixes code (`param.grad = None`) blocking Megatron-FSDP from working with full-iteration CUDA graph replay. Adds an argument / config `cuda_graph_mode` to explicitly tell Megatron-FSDP to use graph-safe behavior when turned on.
  - Some modifications (such as `n_tensor = torch.empty(1, dtype=torch.int64, device=dev).fill_(n)`) were downstreamed from https://github.com/NVIDIA/Megatron-LM/pull/2690
- A further commit extends this PR to work with partial CG as well: https://github.com/cspades/Megatron-LM/tree/xren/nemotron_fsdp_partcg_mambashard
  - @buptzyb has a TE partial CG working after some TE changes:
    - TE: https://github.com/NVIDIA/TransformerEngine/pull/2831
    - MLM: https://github.com/NVIDIA/Megatron-LM/pull/4232

TODO(@cspades): UPDATE DOCUMENTATION, but perhaps after the TE-based partial CG is merged as well so I have the full support matrix at hand.

## Usage

```
# MLM CLI
--cuda-graph-impl full_iteration

# Megatron-FSDP uses CG-safe code.
DDPConfig.megatron_fsdp_cuda_graph_mode=True

# Full-iteration CG only, since this AG is triggered by the DistOpt and not captured by CG, causing an error on wait().
DDPConfig.fsdp_all_gather_in_start_param_sync=False
```

- Refer to `examples/megatron_fsdp/train_llama3_8b_fsdp_h100_fp8.sh`. Assertion guard forces the user to turn on `--use-precision-aware-optimizer` or match main parameter / gradient precision when using full-iter CG.
- Megatron CUDA Graph Current State
  - Local Full-Iteration
    - Works with dense models.
    - For MoE models, requires paged stashing.
  - Local Layer-wise
    - Works with dense models.
    - For MoE models, requires paged stashing.
  - TransformerEngine Partial CG
    - Requires PR to TE: https://github.com/NVIDIA/TransformerEngine/pull/2831
    - Works with dense and MoE models.
  - Local Partial CG
    - Requires `HybridModel`.
    - Works with dense and MoE models.

## Bug

- Megatron-FSDP was not compatible with CUDA graph replay, because we dereference one of the pointers required for the optimizer step after graph capture. **_We need to preserve the `param.(decoupled_)grad` pointer to access the output gradient shards produced by the CUDA graph replay._**
  - `finalize_model_grads() -> update_main_grads() -> setattr(param, "[decoupled_]grad", grad)` installs the sharded gradient.
  - `zero_grad()` dereferences `param.grad` and `param.decoupled_grad`.
  - During replay, the `param.grad` pointer is not re-attached to the sharded gradient, and `optimizer.step()` fails.

## Testing

- @rapatel has proof-of-convergence and increased performance on MLPerf benchmarks.
- Torch-native full-iteration CUDA graph for `fully_shard`, and an E2E `pretrain` test for Megatron-FSDP in MLM. (Couldn't get CUDA graph working with lower-level entrypoints, too many conditionals and training loop logic to port into the unit test.)

## Details

- [`FusedAdam(capturable=True)`](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/optimizers/fused_adam.py) is **_required_** for CUDA graphability, because `Adam.zero_grad()` and most Torch-native optimizers dereference the gradient, i.e. `param.grad = None`, while `FusedAdam.zero_grad(set_to_none=False)` preserves the gradient buffer.
- Recommended CUDA-graphable paths where Megatron-FSDP does not allocate extra memory overhead are:
  - `--use-precision-aware-optimizer` / `megatron_fsdp_use_decoupled_grad=True` - `param.decoupled_grad` is always a view of Megatron-FSDP's sharded main gradient accumulation buffer, which is persistent.
  - `megatron_fsdp_main_params_dtype == megatron_fsdp_main_grads_dtype` - `param.grad` does not require `.to(param.dtype)` casting that allocates unwanted memory overhead.

#### View-Checking the Gradient Shard

- When we are not using the precision-aware optimizer and the main weight precision doesn't match the main gradient precision, we end up with a casted copy of the gradient shard. In this situation, we should deallocate / dereference `param.grad` and this combination of arguments will not be compatible with the full CUDA graph.
```
(Pdb) optimizer_grad
tensor([0., 0., 0.,  ..., 0., 0., 0.], device='cuda:0', dtype=torch.bfloat16)
(Pdb) param.dtype
torch.float32
(Pdb) id(optimizer_grad)
140718698403376
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2982)update_main_grads()
-> if group.main_weight_buffer is not None and not self.use_decoupled_grad:
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2985)update_main_grads()
-> optimizer_grad = optimizer_grad.to(param.dtype)
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2987)update_main_grads()
-> if name not in self.dist_main_grad:

# CASTED GRADIENT / DIFFERENT ADDRESS
(Pdb) id(optimizer_grad)
140718698869968
# NOT A VIEW
(Pdb) optimizer_grad._base
```
- Otherwise, if we are using the precision-aware optimizer, or set `--megatron-fsdp-main-params-dtype == --megatron-fsdp-main-grads-dtype`, then we have a view of the sharded gradient buffer, do not need to dereference the view, and support CUDA graph-ability!
```
(Pdb) optimizer_grad
tensor([0., 0., 0.,  ..., 0., 0., 0.], device='cuda:0')
(Pdb) param.dtype
torch.float32
(Pdb) id(optimizer_grad)
140718436766800
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2982)update_main_grads()
-> if group.main_weight_buffer is not None and not self.use_decoupled_grad:
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2985)update_main_grads()
-> optimizer_grad = optimizer_grad.to(param.dtype)
(Pdb) n
> /opt/megatron-lm/megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py(2987)update_main_grads()
-> if name not in self.dist_main_grad:

# NO-OP / SAME ADDRESS
(Pdb) id(optimizer_grad)
140718436766800
# VIEW OF MFSDP MAIN GRAD BUFFER
(Pdb) optimizer_grad._base
tensor([ 0.0000,  0.0000,  0.0000,  ..., -0.4228, -3.4517, -1.0283],
       device='cuda:0')
```

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 1]
```diff
@@ -2981,7 +3001,7 @@ def update_main_grads(self):
                 setattr(param, "decoupled_grad", grad)
             else:
                 # Attach the gradient to the optimizer parameter.
-                setattr(param, "grad", grad.to(param.dtype) if grad is not None else None)
+                setattr(param, "grad", grad)
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2997]: 90% sure this lower cast is legacy code. Originally this is where we did the cast, but a past precision-aware optimizer PR added a 2nd cast above. Let's just have 1 cast. 👍🏻

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 2]
```diff
@@ -2777,14 +2777,39 @@ def zero_grad(self):
...
+                and param.grad._local_tensor._base is None
+            ):
+                # NOTE: Only when param.dtype != group.main_grad_buffer.dtype.
+                param.grad = None
+            if (
+                hasattr(param, "decoupled_grad")
+                and isinstance(param.decoupled_grad, DTensor)
+                and param.decoupled_grad._local_tensor._base is None
+            ):
+                # NOTE: Decoupled gradients should always be a view of Megatron-FSDP buffers.
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: The newly added code appears to have implemented stricter checks. What is the purpose of this? Will it help resolve full CG compatibility issues?

**Review by shjwudp — APPROVED** [review]
I understand that removing type casting is key to ensuring compatibility with full CG. You've already tested the compatibility between decoupled gradients and full CG—correct?

To avoid blocking MLPerf, I’m approving this PR. However, I suggest we clearly document the conditions required for full CG compatibility and avoid introducing unrelated code changes, as they could cause confusion when we revisit this fix later.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 2 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: This is actually completely up for debate during review, it's not a major blocker either so we can discuss this now. Also, I am still adding more tests and will add documentation when I test full-iteration and partial CG.

TL;DR We have to choose one of the following solutions:

- Limited CUDA graph support (will not work for `param.dtype != grad.dtype`), and the exact same memory utilization as before.
- Full CUDA graph support, but increased sharded memory utilization for certain cases.

**_The only requirement for full-iteration CUDA graph is that we do not set `param.grad = None`_**. However, FSDP dereferences the gradient sometimes:

- If `param.grad` is a view of our sharded buffer, there is no need to deallocate / dereference this variable.
- If `param.grad` is not a view of our sharded buffer, deallocating / dereferencing this variable can save memory, because it is a casted-copy of our sharded buffer.

The logic for `decoupled_grad` is exactly the same. The main difference is that we should never need to run this code for the decoupled gradient, because we never have `decoupled_grad.to(param.dtype)`, and `FusedAdam(use_decoupled_grad=True)` allows for mixed-precision optimization steps.

The complex logic here dereferences this variable when there is memory to reclaim, so we do not persistently hold 2 copies of the sharded gradient that will raise memory overhead.

An alternative is that we can delete all of the `param.grad = None` code, where:

- If it is a view of the sharded buffer, we will zero out the buffer below.
- If it is not a view of the sharded buffer, then `grad.to(param.dtype)` will overwrite this buffer's values during CUDA graph replay.
  - This is a 2nd allocation of the sharded gradient, which increases memory utilization.

In this case, we can support CUDA graph-ability for all cases, but Megatron-FSDP will use more memory in specific cases.

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Would it make sense to introduce CUDA Graph compatibility as an optional path in Megatron-FSDP? This could help us avoid impacting the default execution while still enabling experimentation and performance gains where applicable.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Done, new config called `megatron_fsdp_cuda_graph_mode` and we should inform MLPerf the next time they bump MLM they will need to add this DDPConfig.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py [hunk 3]
```diff
@@ -153,13 +151,13 @@ class DistributedDataParallelConfig:
...
-    def __post_init__(self):
-        import os
-
-        """Check the validity of the config."""
-        if self.nccl_ub and not is_torch_min_version("2.11.0a0"):
-            if 'expandable_segments:True' in os.getenv('PYTORCH_CUDA_ALLOC_CONF', '').split(','):
-                raise ValueError(
-                    "PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True is currently not supported "
-                    "with nccl_ub due to compatibility issue with torch.cuda.MemPool API."
-                )
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py:176]: We don't need NCCL UBR checks in the PyPI package version since we do not currently have a support plan for NCCL UBR. (Requires the user to call manual registration after the first step.) @ahmadki can pipeclean it later as well.

Importing MCore here breaks PyPI publishing, we need to revert this.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 4]
```diff
@@ -2777,14 +2777,31 @@ def zero_grad(self):
...
+        distributed optimizer parameter precision, the gradient shard is a view of
+        the Megatron-FSDP sharded gradient buffer. If not, then not dereferencing
+        this gradient shard will increase memory utilization as this gradient is a
+        persistent casted-copy of the accumulated gradient.
+        """
+        if not self.ddp_config.megatron_fsdp_cuda_graph_mode:
+            # Dereference the sharded gradient to reclaim memory
+            # unless a full-iteration CUDA graph is utilized.
+            for name, param in self.optimizer_named_parameters:
+                param.grad = None
```
Comment (galv) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2798]: FYI, once we merge this https://github.com/pytorch/pytorch/pull/174649, you can automatically detect whether your cuda graph references a deleted tensor, rather than having to hunt for them manually. I'm not sure if finding the offending `param.grad = None` calls was hard for you, but if it was, here you go.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2798]: Oh cool! Yes, that would have made it a bit easier! (@nanz-nv helped me hunt this down.)

File: megatron/training/training.py [hunk 5]
```diff
@@ -1717,6 +1717,16 @@ def get_megatron_ddp_config(args: argparse.Namespace) -> DistributedDataParallel
         kwargs["megatron_fsdp_main_grads_dtype"] = args.megatron_fsdp_main_grads_dtype
         kwargs["megatron_fsdp_grad_comm_dtype"] = args.megatron_fsdp_grad_comm_dtype
         kwargs["megatron_fsdp_use_decoupled_grad"] = args.use_precision_aware_optimizer
+        if args.use_megatron_fsdp and args.cuda_graph_impl != "none":
```
Comment (maanug-nv) [inline megatron/training/training.py:1924]: don't really care if this check lives here or in validate_args right now, but we should move it to config container validate after we can merge #4618.

**Review by maanug-nv — APPROVED** [review]

**Comment by rapatel** [comment]
@cspades How much effort would it be to create a PR of this into dev branch? Reason I ask is because next week we're planning to unfreeze the mlperf repo back to mbridge:main + mcore:dev. And I'd like to enable e2e runs in the mlperf nightly with fullCG+fsdp configs. To do that, I'd need this PR (or temporarily cherry pick what's needed from here).

**Comment by cspades** [comment]
> @cspades How much effort would it be to create a PR of this into dev branch? Reason I ask is because next week we're planning to unfreeze the mlperf repo back to mbridge:main + mcore:dev. And I'd like to enable e2e runs in the mlperf nightly with fullCG+fsdp configs. To do that, I'd need this PR (or temporarily cherry pick what's needed from here).

@rapatel Not difficult at all! I'll do it soon. I think this PR is still ready-to-merge, I'll ping people to approve.

**Review by jaredcasper — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by youngeunkwon0405 — APPROVED** [review]
